lib/lambda_ethereum_consensus/p2p/block_downloader.ex

```diff
@@ -15,100 +17,68 @@ defmodule LambdaEthereumConsensus.P2P.BlockDownloader do
          :ok <- verify_batch(blocks, slot, count) do
       tags = %{result: "success", type: "by_slot", reason: "success"}
       :telemetry.execute([:network, :request], %{blocks: count}, tags)
       {:ok, blocks}
     else
-      {:error, reason} when retries > 0 ->
+      {:error, reason} ->
         tags = %{type: "by_slot", reason: parse_reason(reason)}
         P2P.Peerbook.penalize_peer(peer_id)
-        :telemetry.execute([:network, :request], %{blocks: 0}, Map.put(tags, :result, "retry"))
-        Logger.debug("Retrying request for block", slot: slot)
-        request_blocks_by_slot(slot, count, retries - 1)
 
-      {:error, reason} when retries == 0 ->
-        P2P.Peerbook.penalize_peer(peer_id)
-        tags = %{type: "by_slot", reason: parse_reason(reason)}
-        :telemetry.execute([:network, :request], %{blocks: 0}, Map.put(tags, :result, "error"))
-        {:error, reason}
+        if retries > 0 do
+          :telemetry.execute([:network, :request], %{blocks: 0}, Map.put(tags, :result, "retry"))
+          Logger.debug("Retrying request for #{count} blocks", slot: slot)
+          request_blocks_by_slot(slot, count, retries - 1)
+        else
+          :telemetry.execute([:network, :request], %{blocks: 0}, Map.put(tags, :result, "error"))
+          {:error, reason}
+        end
     end
   end
 
   @spec request_block_by_root(Types.root(), integer()) ::
-          {:ok, Types.SignedBeaconBlock.t()} | {:error, binary()}
+          {:ok, SignedBeaconBlock.t()} | {:error, binary()}
   def request_block_by_root(root, retries \\ @default_retries) do
     with {:ok, [block]} <- request_blocks_by_root([root], retries) do
       {:ok, block}
     end
   end
 
   @spec request_blocks_by_root([Types.root()], integer()) ::
-          {:ok, [Types.SignedBeaconBlock.t()]} | {:error, binary()}
+          {:ok, [SignedBeaconBlock.t()]} | {:error, binary()}
   def request_blocks_by_root(roots, retries \\ @default_retries)
 
-  def request_blocks_by_root([], _retries) do
-    {:ok, []}
-  end
+  def request_blocks_by_root([], _retries), do: {:ok, []}
 
   def request_blocks_by_root(roots, retries) do
     Logger.debug("Requesting block for roots #{Enum.map_join(roots, ", ", &Base.encode16/1)}")
 
     peer_id = get_some_peer()
 
-    # TODO ssz encode array of roots
-    # {:ok, encoded_payload} = payload |> Ssz.to_ssz()
-    encoded_payload = Enum.join(roots)
-
-    size_header =
-      encoded_payload
-      |> byte_size()
-      |> P2P.Utils.encode_varint()
-
-    {:ok, compressed_payload} = encoded_payload |> Snappy.compress()
-
-    with {:ok, response_chunk} <-
-           Libp2pPort.send_request(
-             peer_id,
-             @blocks_by_root_protocol_id,
-             size_header <> compressed_payload
-           ),
-         {:ok, chunks} <- parse_response(response_chunk),
-         {:ok, blocks} <- decode_chunks(chunks) do
+    schema = {:list, TypeAliases.root(), ChainSpec.get("MAX_REQUEST_BLOCKS")}
```

Comment:
Maybe we should have a type alias of this? e.g. `TypeAliases.block_root_list`? We are using this exact same thing in  `BeaconBlockByRootRequestContainer`. If we can't use that container directly, we should at least alias this list.

Reply:
I changed it so it now uses the struct instead of the raw schema.